docs: Add user guide for GPU annotation and resource requests in vertices

sapkota-aayush opened the pull request:
This PR adds comprehensive documentation for configuring GPU resources in Numaflow vertices, including:
1. How to request GPU resources in both Pipeline and MonoVertex specs
3. How and why to add annotations for GPU scheduling or monitoring
4. Node selector and security context examples for GPU workloads
5. Clarification on Kubernetes requirements for GPU resource requests and limits
6. A minimal example for Dynamic Resource Allocation (DRA) with GPUs
7. References to best practices and relevant internal/external documentation
The new guide appears in the User Guide > Reference section of the documentation UI.

Closes #1057

sapkota-aayush commented:
@vigith @KeranYang @syayi 
I only updated documentation in this PR and did not change any application code or tests. The CI failure seems unrelated to my changes.
Do I need to take any action to resolve this, or can it be handled/overridden by a maintainer? Please advise.
Thank you!

vigith commented:
I have rerun the failed test, it seems to be an issue with the infra.

vigith commented on docs/user-guide/reference/gpu-annotation.md:
````diff
@@ -0,0 +1,191 @@
+
+For advanced or dynamic GPU scheduling, Numaflow supports [Dynamic Resource Allocation (DRA)](./configuration/dra.md). This allows for more flexible and powerful GPU resource management.
+
+**Minimal Example:**
+
+```yaml
+apiVersion: resource.k8s.io/v1alpha3
+kind: ResourceClaimTemplate
+metadata:
+  name: my-gpu
+spec:
+  spec:
+    devices:
+      requests:
+        - name: gpu
+          deviceClassName: gpu.nvidia.com
+---
+apiVersion: numaflow.numaproj.io/v1alpha1
+kind: Pipeline
+metadata:
+  name: dra-gpu-pipeline
+spec:
+  vertices:
+    - name: inference
+      udf:
+        container:
+          image: my-image
+          resources:
+            claims:
+              - name: gpu
+      resourceClaims:
+        - name: gpu
+          resourceClaimTemplateName: my-gpu
+```
````
isn't this a duplicate of the DRA doc?

vigith commented on docs/user-guide/reference/gpu-annotation.md:
```diff
@@ -0,0 +1,191 @@
+## Using GPU Resources in a Numaflow Vertex
+
+### What is a Vertex?
+
+In Numaflow, a **vertex** is a core component of a pipeline. Each vertex represents a processing step, such as:
+- Reading data from a source
+- Transforming data using a user-defined function (UDF)
+- Aggregating or reducing data
+- Writing data to a sink
+
+By connecting multiple vertices, you can build complex data processing pipelines for streaming or batch workloads.
```
let's point to vertex.md to explain what vertex is, else the doc will become very verbose.

vigith commented on docs/user-guide/reference/gpu-annotation.md:
```diff
@@ -0,0 +1,191 @@
+> **Note:**
+> All the guidance in this document (GPU resource requests, annotations, node selectors, etc.) applies to both `Pipeline` and `MonoVertex` specifications in Numaflow.
+
+### Why Use GPU Resources?
+
+Some data processing tasks are highly computational and can be executed significantly faster on a **GPU (Graphics Processing Unit)** than on a CPU. Common scenarios where GPU acceleration is beneficial include:
+- Running machine learning or deep learning models
+- Processing large images or videos
+- Performing scientific or mathematical computations
```
you can remove this too. perhaps link to a wikipedia page?

vigith commented on docs/user-guide/reference/gpu-annotation.md:
```diff
@@ -0,0 +1,191 @@
+- Processing large images or videos
+- Performing scientific or mathematical computations
+
+## Prerequisites
```
you can reduce the content a lot in this section.

vigith reviewed (changes requested):
could you please reduce the verbosity a bit? users trying to enable GPU knows about prerequisites. add only those that really matters for enabling GPU with Numaflow

sapkota-aayush commented:
> @sapkota-aayush thanks, I have two comments where I suggested to remove some lines, could you take a look? I resolved others.

I’ve removed the extra line as you suggested to further reduce verbosity.
Regarding the commented-out securityContext YAML block: I’ve kept it minimal and concise, but let me know if you’d prefer it removed entirely. Happy to update as needed!
Thanks for your feedback!

KeranYang commented:
@sapkota-aayush sorry, my suggestion was to remove `### What is a Vertex?` and `### GPU Resources` sections and use a single sentence of `GPU resources can be configured on a Pipeline Vertex or a MonoVertex.`. If users visit this page, they should already know what Vertex and GPU is.

Please remove `securityContext` from the example spec. In general, we put minimal spec in examples. User can be very creative when configuring other fields and we don't make assumptions on what they need.

sapkota-aayush commented:
> @sapkota-aayush sorry, my suggestion was to remove `### What is a Vertex?` and `### GPU Resources` sections and use a single sentence of `GPU resources can be configured on a Pipeline Vertex or a MonoVertex.`. If users visit this page, they should already know what Vertex and GPU is.
> 
> Please remove `securityContext` from the example spec. In general, we put minimal spec in examples. User can be very creative when configuring other fields and we don't make assumptions on what they need.

All requested changes have been made:
Removed the “What is a Vertex?” and “GPU Resources” sections, and replaced them with a single concise sentence at the top.
Removed the commented-out securityContext block from the YAML example.

sapkota-aayush commented:
> @sapkota-aayush one last comment - please have a title just to be consistent with other docs. `# Configuring GPU`

Added the title "# Configuring GPU" at the top of the document for consistency with other docs. Thanks for the suggestion!
